enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,497 @@
+local changes and rebooting. In a single-node deployment, there is
+nowhere for drained pods to be rescheduled, so when the
+high-availability mode is none this step needs to be skipped. Ideally
+we will be able to rely on the [graceful shutdown support in
+kubelet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown)
+to stop all workloads safely as part of the reboot.
+
+#### cluster-authentication-operator
+
+By default, the `cluster-authentication-operator` will not deploy
+oauth-apiserver and oauth-server without a minimum of 3 control plane
+nodes. This can be change by enabling
+`useUnsupportedUnsafeNonHANonProductionUnstableOAuthServer`. The
+operator will need to be updated to honor the new capabilities API
+instead. Health checks for these components will also have to be
+updated to allow a single replica.
+
+#### cluster-etcd-operator
+
+By default, `cluster-etcd-operator` will not deploy the etcd cluster
+without minimum of 3 master nodes. This can be changed by enabling
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd`. The operator
+will need to be updated to honor the new capabilities API instead.
+
+`etcd-quorum-guard` is managed by the cluster-version-operator, which
+means it has a statically configured Deployment. The operator will
+need to be updated to manage the `etcd-quorum-guard` so that it can
+support changing the replica count or disabling the guard entirely,
+depending on the best approach to honor the high-availability mode
+setting in the capabilities API.
+
+#### cluster-ingress-operator
+
+By default, the `cluster-ingress-operator` deploys the router with 2
+replicas. On a single node one will fail to start and the ingress will
+show as degraded. The operator will need to be updated to honor the
+new capabilities API to change the replica count to 1 when the
+high-availability mode is none.
+
+#### cluster-machine-approver
+
+The operator should disable automatic machine approval. (See risks
+section below.)
+
+#### operator-lifecycle-manager
+
+1. OLM operators may be run on other Kubernetes distributions, so they
+   should not rely on an OpenShift API. We need a way to pass the HA
+   mode flag to those operators. One possible solution is using an enviroment
+   variable but needs to discussed further.
+
+### Risks and Mitigations
+
+*How will security be reviewed and by whom? How will UX be reviewed and by whom?*
+
+#### Control plane rollouts
+
+kube-apiserver is periodically reconfigured by its operator in
+response to changes in cluster state (e.g. weekly rotation of etcd
+encryption keys). When this occurs, the apiserver may be inaccessible
+for up to 2 minutes. Workloads that do not depend on the apiserver
+would be expected to run without issue during this interval. Workloads
+that do depend on apiserver availability would need to be resilient to
+these events. OpenShift core components are already resilient in this
+way.
```

Comment:
(Maybe this closely related to my question on the configuration changes discussion)

This section begs more questions for me than answers - e.g. what do we mean by "rollouts"? Reconfiguration for key rotations I think I get, but are there other examples of "periodically reconfigured"? Inaccessible for up to 2 minutes seems very specific - is this 2 minute timeframe somehow fundamental, or something that can be improved? Can we be more specific about how OpenShift core components are resilient to this, and how other workloads are going to need to be adapted?

Thanks,

Reply:
@deads2k @cgwalters I think these details came from one of you, can you help answer the question?

Reply:
Pinging @deads2k and @cgwalters for help here.

Reply:
As I understand it, the new encryption keys cause a restart, or at least a temporary pause while the new keys are loaded. I guess restarting takes around 2 minutes? @deads2k is that right?